### PR TITLE
Field and schema extensions

### LINKED ISSUE #1791: Can protoc-gen-swagger hide the specified fields?
## 🚀 Feature

I want to hide certain fields in the json generated by protoc-gen-swagger

### PR DESCRIPTION
**Pull request by james-o-johnstone**
#### References to other Issues or PRs
Fixes  #1791

#### Have you read the [Contributing Guidelines](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/CONTRIBUTING.md)?
Yes

#### Brief description of what is fixed or changed
Add extensions to field and schema annotations.

#### Other comments
Support for extensions was added in https://github.com/grpc-ecosystem/grpc-gateway/pull/1033 but not for field/schema annotations.

### CODE REVIEW

File: protoc-gen-openapiv2/internal/genopenapi/template.go
Comment (johanbrandhorst) [inline protoc-gen-openapiv2/internal/genopenapi/template.go:415]: Would it be worth checking that the key of the extensions start with `x-` as specified in the swagger spec: https://swagger.io/docs/specification/2-0/swagger-extensions/?

File: protoc-gen-openapiv2/options/openapiv2.proto
Comment (johanbrandhorst) [inline protoc-gen-openapiv2/options/openapiv2.proto:388]: Which section within the swagger spec does this map to: https://swagger.io/docs/specification/2-0/swagger-extensions/? Just want to make sure we're not adding these somewhere they don't belong.

File: protoc-gen-openapiv2/options/openapiv2.proto
Comment (johanbrandhorst) [inline protoc-gen-openapiv2/options/openapiv2.proto:509]: Which section within the swagger spec does this map to: https://swagger.io/docs/specification/2-0/swagger-extensions/? Just want to make sure we're not adding these somewhere they don't belong.

**Review by johanbrandhorst — COMMENTED** [review]
This looks great! I had some small questions. To fix the CLA bot you'll need to force push over any old commits on an account not linked to github.

File: protoc-gen-openapiv2/options/openapiv2.proto
Comment (james-o-johnstone) [reply protoc-gen-openapiv2/options/openapiv2.proto:388]: If I'm understanding correctly, this is used for [`MessageOptions`](https://github.com/grpc-ecosystem/grpc-gateway/blob/185126621f298395754503abc6460e5d9e1f62b9/protoc-gen-openapiv2/options/annotations.proto#L29) annotations and so could be either operation parameters or responses (see https://github.com/grpc-ecosystem/grpc-gateway/blob/185126621f298395754503abc6460e5d9e1f62b9/protoc-gen-openapiv2/options/openapiv2.proto#L230) in the swagger schema.

This allows for annotations like this, e.g.:

```
message AMessage {
  option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_schema) = {
    extensions: {
      key: "x-something-something";
      value {
        string_value: "yadda";
      }
    }
  }
};
```

File: protoc-gen-openapiv2/options/openapiv2.proto
Comment (james-o-johnstone) [reply protoc-gen-openapiv2/options/openapiv2.proto:509]: If I'm understanding correctly, this is used for [`FieldOptions`](https://github.com/grpc-ecosystem/grpc-gateway/blob/185126621f298395754503abc6460e5d9e1f62b9/protoc-gen-openapiv2/options/annotations.proto#L29) annotations, I think this maps to the swagger schema object which supports extensions according to: https://github.com/OAI/OpenAPI-Specification/blob/3.0.0/versions/2.0.md#patterned-objects-9. Example:

```
message AMessage {
  string id = 1 [
    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
      extensions: {
        key: "x-something-something";
        value {
          string_value: "yadda";
        }
      }
    }
  ]
};
```

However, in revisiting this I've realised that the `JSONSchema` defined here is also part of the `Schema` type defined above, so it looks like it would be possible to only add the `extensions` field in this `JSONSchema` message and use in schema annotations like this:

```
message AMessage {
  option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_schema) = {
    json_schema: {
      extensions: {
        key: "x-something-something";
        value {
          string_value: "yadda";
        }
      }
    }
  };
}
```

File: protoc-gen-openapiv2/internal/genopenapi/template.go
Comment (james-o-johnstone) [reply protoc-gen-openapiv2/internal/genopenapi/template.go:415]: Good point, fortunately this is already handled in the [`processExtensions` ](https://github.com/grpc-ecosystem/grpc-gateway/blob/b5cd496c648a3394eacb65d9e2f6f07adb2449df/protoc-gen-openapiv2/internal/genopenapi/template.go#L1538) function called a bit further up.

File: protoc-gen-openapiv2/options/openapiv2.proto
Comment (johanbrandhorst) [reply protoc-gen-openapiv2/options/openapiv2.proto:509]: That sounds reasonable to me

**Comment by drewwells** [comment]
Looking forward to this getting merged. Will it happen soon?

**Comment by johanbrandhorst** [comment]
Unfortunately we're still waiting for CLA approval here. Note that the Google CLA bot has been acting up a bit lately so I may need to manually override it, as long as @james-o-johnstone can provide some evidence that he's signed it I'm happy.

**Comment by drewwells** [comment]
> Unfortunately we're still waiting for CLA approval here. Note that the Google CLA bot has been acting up a bit lately so I may need to manually override it, as long as @james-o-johnstone can provide some evidence that he's signed it I'm happy.

I don’t have context on these changes. If they are all good, I can definitely open a new PR and sign the CLA. Pretty sure I already signed the CLA

**Comment by johanbrandhorst** [comment]
I wouldn't be comfortable with simply copy-pasting the contents of the PR without explicit consent from @james-o-johnstone. If they're unable to sign the CLA but can consent to use copying the contribution we could make it happen in a separate PR.

**Comment by james-o-johnstone** [comment]
Hey, sorry I'm still waiting on the contributor group to be set up at my company before I can sign the CLA, I will chase this up today.

**Comment by drewwells** [comment]
> Hey, sorry I'm still waiting on the contributor group to be set up at my company before I can sign the CLA, I will chase this up today.

Thanks @james-o-johnstone

**Comment by james-o-johnstone** [comment]
Hey, sorry for the delay on this. This is still ticking along internally, CLA has now been reviewed by legal team and hopefully the wait won't be too much longer before I can sign the CLA. Thanks for your patience.

**Comment by johanbrandhorst** [comment]
Thanks for the update James, in the meantime, maybe you want to try rebasing on master?

**Comment by C0mbatwombat** [comment]
Thanks James for the hard work, I am waiting eagerly for this change :+1:

**Comment by johanbrandhorst** [comment]
There is now an alternative implementation to hiding fields/RPCs in https://github.com/grpc-ecosystem/grpc-gateway/pull/2578. We probably only want to merge one, so would interested parties weigh in on which they would prefer?

**Comment by james-o-johnstone** [comment]
@googlebot I signed it!

**Review by johanbrandhorst — APPROVED** [review]
LGTM 🎉

**Comment by johanbrandhorst** [comment]
Ah one final thing I would ask, could you add something to document this new feature? The docs for the openapi options are found here: https://github.com/grpc-ecosystem/grpc-gateway/blob/master/docs/docs/mapping/customizing_openapi_output.md.

**Review by johanbrandhorst — APPROVED** [review]

**Comment by johanbrandhorst** [comment]
Thanks for your contribution! I'm extremely pleased to have this merged finally :).

**Pull request merged**
